lib/src/time_util.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides support for parsing and matching date ranges.
+
+use chrono::{DateTime, Local, TimeZone, Utc};
+use chrono_english::{parse_date_string, DateError, Dialect};
+use thiserror::Error;
+
+use crate::backend::Timestamp;
+
+/// Context needed to create a DatePattern during revset evaluation.
+#[derive(Clone, Debug)]
+pub enum DatePatternContext {
+    /// Interpret date patterns using the local machine's time zone
+    Local(DateTime<Local>),
+    /// Interpret date patterns using UTC
+    Utc(DateTime<Utc>),
+}
+
+impl DatePatternContext {
+    /// Parses a DatePattern from the given string and kind.
+    pub fn to_date_pattern(
```

Comment:
nit: `to_` sounds a bit odd. Maybe this can be renamed to `parse_relative`/`_with` or just `parse`?